### DISCUSSION TITLE
tox4: Executable name, why naming it tox4?

### DISCUSSION THREAD

**Discussion by JulienPalard** [tox-dev/tox#1856] [Q&A] [answered]
I understand the rewrite has temporarily been renamed `tox4` to ease its test "along with tox 3", but as `pip` uninstalls tox 3 while installing tox4, it's not as helpfull as I expected, or am I missing something?

**Comment by gaborbernat** [comment, marked as answer]
The idea is that you install tox 4 into a dedicated virtual environment, not into a global python environment. In which case, things work as expected. Otherwise, we'd have to rename all source files and publish under different names, which is a heavier lift.

**Reply by JulienPalard** [reply]
I did installed tox 4 in a dedicated venv, still can't have tox and tox4 together, so I still don't see the benefit:

```
$ pip install tox
$ pip install .
Installing collected packages: colorama, tox
  Attempting uninstall: tox
    Found existing installation: tox 3.21.2
    Uninstalling tox-3.21.2:
      Successfully uninstalled tox-3.21.2
Successfully installed colorama-0.4.4 tox-4.0.0a5.dev1+g5f172bc
```

Oh I think I get it, you're installing your tox globally and tox4 in the venv?

**Reply by gaborbernat** [reply]
I'm installing both in their dedicated virtual environments. When I mean side by side I mean executable side by side. So you can invoke either one by just typing tox or tox4.

**Reply by gaborbernat** [reply]
Think of it has linking the virtual environment executable to `/usr/bin/local/tox` and `/usr/bin/local/tox4` and then being able to call either `tox` or `tox4`. On Windows I don't do the linking, just put both script folders on the PATH.

**Reply by JulienPalard** [reply]
Ooooh ok the idea is to add multiple venv's bin folders to your PATH? A bit like having multiple venvs activated at once... If I understand correctly you're having a 3rd venv "activated" in front of them all with no tox installed in them, so the "3rd venv" (let's call it the "project venv"?) gets `pip`, `python`, ...?

**Reply by gaborbernat** [reply]
I personally never activate my virtual environments. IMHO it's a bad idea 🤷🏻  and makes you question yourself where things will execute. I prefer to type out the relative paths, which IMHO is the sane way. You're explicit about where the command will get executed. E.g. ``venv/bin/pip install whatever`` or ``venv/bin/python``.  You know that mantra from the zen of python: `Explicit is better than implicit.`. Tools, such as tox/black/pre-commit, are the exception to typing out the full/relative path because the fact they are written in Python is an implementation detail. But I do make sure the only thing you get onto the path out of their virtual environment is the console entry point of them, https://pipxproject.github.io/pipx/installation/ makes this really easy.

**Reply by JulienPalard** [reply]
OK it starts to make sense, thanks for clarifying, really appreciated!
